### PR TITLE
CI updates

### PR DESCRIPTION
**Pull request by hashhar**
## Description

- Update GitHub action versions
- Add pip caching for CI workflows, release workflow doesn't use caching still just for safety/clean build reasons

## Release notes

(x) This is not user-visible or docs only and no release notes are required.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -20,12 +20,13 @@ jobs:
...
       - name: "Checkout the source code"
-        uses: actions/checkout@v3
+        uses: actions/checkout@v4
 
       - name: "Install Python"
-        uses: actions/setup-python@v4
+        uses: actions/setup-python@v5
         with:
           python-version: "3.11"
+          cache: "pip"
```
Comment (nineinchnick) [inline .github/workflows/ci.yml:29]: I'm not sure if caching can be used without `requirements.txt`, maybe it's not worth anyway?

Comment (hashhar) [reply .github/workflows/ci.yml:29]: it works with setup.py too but I forgot to add the params. Thanks.

Comment (hashhar) [reply .github/workflows/ci.yml:29]: Eventually we plan to move to pyproject.toml now that setup.py is going to be deprecated soon.

Comment (hashhar) [reply .github/workflows/ci.yml:29]: saves ~1m per job for total of 11 mins across a run. I'll take what I can get.

**Review by hovaesco — APPROVED** [review]

**Pull request merged**